File: data-inspector/src/main/java/com/example/datainspector/provider/JpaDataSourceProvider.java
package com.example.datainspector.provider;

import com.example.datainspector.model.DataSourceInfo;
import com.example.datainspector.model.QueryResult;
import com.example.datainspector.spi.DataSourceProvider;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.Query;
import jakarta.persistence.TypedQuery;
import jakarta.persistence.metamodel.EntityType;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Provides JPA entity data sources for SQL databases
 * Discovers all JPA entities and allows querying them
 */
@Slf4j
@Component
public class JpaDataSourceProvider implements DataSourceProvider {

    private final EntityManagerFactory entityManagerFactory;
    private final ObjectMapper objectMapper;

    public JpaDataSourceProvider(Optional<EntityManagerFactory> entityManagerFactory, ObjectMapper objectMapper) {
        this.entityManagerFactory = entityManagerFactory.orElse(null);
        this.objectMapper = objectMapper;
        if (entityManagerFactory.isPresent()) {
            log.info("JPA data source provider enabled");
        }
    }

    @Override
    public List<DataSourceInfo> discoverDataSources() {
        if (entityManagerFactory == null) {
            return Collections.emptyList();
        }

        try {
            Set<EntityType<?>> entities = entityManagerFactory.getMetamodel().getEntities();
            EntityManager em = entityManagerFactory.createEntityManager();

            List<DataSourceInfo> dataSources = new ArrayList<>();

            // Add database overview
            dataSources.add(DataSourceInfo.builder()
                    .id("jpa:overview")
                    .name("Database Overview")
                    .type("jpa-overview")
                    .description("Overview of all JPA entities and database statistics")
                    .size(entities.size())
                    .queryable(true)
                    .build());

            // Add each entity as a data source
            for (EntityType<?> entity : entities) {
                String entityName = entity.getName();
                Class<?> javaType = entity.getJavaType();

                try {
                    long count = getEntityCount(em, entityName);

                    dataSources.add(DataSourceInfo.builder()
                            .id("jpa:entity:" + entityName)
                            .name("Entity: " + entityName)
                            .type("jpa-entity")
                            .description("JPA Entity: " + javaType.getName())
                            .size(count)
                            .queryable(true)
                            .metadata(Map.of(
                                    "entityName", entityName,
                                    "javaType", javaType.getName(),
                                    "tableName", getTableName(entity)
                            ))
                            .build());
                } catch (Exception e) {
                    log.debug("Could not get count for entity {}: {}", entityName, e.getMessage());
                }
            }

            em.close();
            return dataSources;
        } catch (Exception e) {
            log.warn("JPA is not available or no entities found: {}", e.getMessage());
            return Collections.emptyList();
        }
    }

    @Override
    public QueryResult query(String dataSourceId, Map<String, Object> filters, int limit, int offset) {
        if (dataSourceId.equals("jpa:overview")) {
            return queryOverview();
        } else if (dataSourceId.startsWith("jpa:entity:")) {
            String entityName = dataSourceId.substring("jpa:entity:".length());
            return queryEntity(entityName, filters, limit, offset);
        }

        return QueryResult.builder()
                .dataSourceId(dataSourceId)
                .data(Collections.emptyList())
                .totalCount(0)
                .build();
    }

    @Override
    public boolean supports(String dataSourceId) {
        return dataSourceId != null && dataSourceId.startsWith("jpa:");
    }

    private long getEntityCount(EntityManager em, String entityName) {
        Query query = em.createQuery("SELECT COUNT(e) FROM " + entityName + " e");
        return ((Number) query.getSingleResult()).longValue();
    }

    private String getTableName(EntityType<?> entity) {
        try {
            jakarta.persistence.Table table = entity.getJavaType().getAnnotation(jakarta.persistence.Table.class);
            if (table != null && !table.name().isEmpty()) {
                return table.name();
            }
            return entity.getName();
        } catch (Exception e) {
            return entity.getName();
        }
    }

    private QueryResult queryOverview() {
        EntityManager em = entityManagerFactory.createEntityManager();
        Set<EntityType<?>> entities = entityManagerFactory.getMetamodel().getEntities();

        List<Map<String, Object>> data = new ArrayList<>();

        for (EntityType<?> entity : entities) {
            String entityName = entity.getName();
            try {
                long count = getEntityCount(em, entityName);
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("entityName", entityName);
                row.put("javaType", entity.getJavaType().getName());
                row.put("tableName", getTableName(entity));
                row.put("recordCount", count);
                row.put("attributes", entity.getAttributes().size());
                data.add(row);
            } catch (Exception e) {
                log.debug("Error querying entity {}: {}", entityName, e.getMessage());
            }
        }

        em.close();

        return QueryResult.builder()
                .dataSourceId("jpa:overview")
                .data(data)
                .totalCount(data.size())
                .limit(data.size())
                .offset(0)
                .build();
    }

    private QueryResult queryEntity(String entityName, Map<String, Object> filters, int limit, int offset) {
        EntityManager em = entityManagerFactory.createEntityManager();

        try {
            // Get total count
            long totalCount = getEntityCount(em, entityName);

            // Build query with filters
            StringBuilder jpql = new StringBuilder("SELECT e FROM " + entityName + " e");

            if (filters != null && !filters.isEmpty()) {
                jpql.append(" WHERE ");
                List<String> conditions = new ArrayList<>();
                for (String key : filters.keySet()) {
                    conditions.add("e." + key + " = :" + key);
                }
                jpql.append(String.join(" AND ", conditions));
            }

            TypedQuery<?> query = em.createQuery(jpql.toString(), Object.class);

            // Set filter parameters
            if (filters != null) {
                for (Map.Entry<String, Object> entry : filters.entrySet()) {
                    query.setParameter(entry.getKey(), entry.getValue());
                }
            }

            // Apply pagination
            query.setFirstResult(offset);
            query.setMaxResults(limit);

            List<?> results = query.getResultList();

            // Convert entities to maps
            List<Map<String, Object>> data = results.stream()
                    .map(this::convertEntityToMap)
                    .collect(Collectors.toList());

            return QueryResult.builder()
                    .dataSourceId("jpa:entity:" + entityName)
                    .data(data)
                    .totalCount(totalCount)
                    .limit(limit)
                    .offset(offset)
                    .stats(Map.of("entityName", entityName))
                    .build();
        } catch (Exception e) {
            log.error("Error querying entity {}: {}", entityName, e.getMessage(), e);
            return QueryResult.builder()
                    .dataSourceId("jpa:entity:" + entityName)
                    .data(Collections.emptyList())
                    .totalCount(0)
                    .limit(limit)
                    .offset(offset)
                    .build();
        } finally {
            em.close();
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> convertEntityToMap(Object entity) {
        try {
            // Use ObjectMapper to convert entity to map
            return objectMapper.convertValue(entity, Map.class);
        } catch (Exception e) {
            log.debug("Could not convert entity to map: {}", e.getMessage());
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("type", entity.getClass().getSimpleName());
            map.put("value", entity.toString());
            return map;
        }
    }
}

File: data-inspector-scala/src/main/scala/com/example/datainspector/provider/JpaDataSourceProvider.scala
/*
 * Copyright (c) 2026 Data Inspector Contributors
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.example.datainspector.provider

import com.example.datainspector.model.{DataSourceInfo, QueryResult}
import com.example.datainspector.spi.DataSourceProvider
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.persistence.{EntityManager, EntityManagerFactory, TypedQuery}
import jakarta.persistence.metamodel.EntityType
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.stereotype.Component

import java.util.Optional
import scala.jdk.CollectionConverters._
import scala.util.{Try, Using}

/**
 * Provides JPA entity data sources for SQL databases
 * Discovers all JPA entities and allows querying them
 */
@Component
class JpaDataSourceProvider(
  entityManagerFactoryOpt: Optional[EntityManagerFactory],
  objectMapper: ObjectMapper
) extends DataSourceProvider {

  private val logger: Logger = LoggerFactory.getLogger(classOf[JpaDataSourceProvider])
  private val entityManagerFactory: Option[EntityManagerFactory] =
    if (entityManagerFactoryOpt.isPresent) Some(entityManagerFactoryOpt.get()) else None

  if (entityManagerFactory.isDefined) {
    logger.info("JPA data source provider enabled")
  }

  override def discoverDataSources(): List[DataSourceInfo] = {
    entityManagerFactory match {
      case None => List.empty
      case Some(emf) =>
        Try {
          val entities = emf.getMetamodel.getEntities.asScala.toSet
          val em = emf.createEntityManager()

          try {
            val dataSources = scala.collection.mutable.ListBuffer[DataSourceInfo]()

            // Add database overview
            dataSources += DataSourceInfo(
              id = "jpa:overview",
              name = "Database Overview",
              `type` = "jpa-overview",
              description = "Overview of all JPA entities and database statistics",
              size = entities.size,
              queryable = true
            )

            // Add each entity as a data source
            entities.foreach { entity =>
              val entityName = entity.getName
              val javaType = entity.getJavaType

              Try {
                val count = getEntityCount(em, entityName)

                dataSources += DataSourceInfo(
                  id = s"jpa:entity:$entityName",
                  name = s"Entity: $entityName",
                  `type` = "jpa-entity",
                  description = s"JPA Entity: ${javaType.getName}",
                  size = count,
                  queryable = true,
                  metadata = Map(
                    "entityName" -> entityName,
                    "javaType" -> javaType.getName,
                    "tableName" -> getTableName(entity)
                  )
                )
              }.recover {
                case e: Exception =>
                  logger.debug(s"Could not get count for entity $entityName: ${e.getMessage}")
              }
            }

            dataSources.toList
          } finally {
            em.close()
          }
        }.recover {
          case e: Exception =>
            logger.warn(s"JPA is not available or no entities found: ${e.getMessage}")
            List.empty[DataSourceInfo]
        }.getOrElse(List.empty)
    }
  }

  override def query(dataSourceId: String, filters: Map[String, AnyRef], limit: Int, offset: Int): QueryResult = {
    if (dataSourceId == "jpa:overview") {
      queryOverview()
    } else if (dataSourceId.startsWith("jpa:entity:")) {
      val entityName = dataSourceId.substring("jpa:entity:".length)
      queryEntity(entityName, filters, limit, offset)
    } else {
      QueryResult(
        dataSourceId = dataSourceId,
        data = List.empty,
        totalCount = 0,
        limit = limit,
        offset = offset
      )
    }
  }

  override def supports(dataSourceId: String): Boolean = {
    dataSourceId != null && dataSourceId.startsWith("jpa:")
  }

  private def getEntityCount(em: EntityManager, entityName: String): Long = {
    val query = em.createQuery(s"SELECT COUNT(e) FROM $entityName e")
    query.getSingleResult.asInstanceOf[Number].longValue()
  }

  private def getTableName(entity: EntityType[_]): String = {
    Try {
      val table = entity.getJavaType.getAnnotation(classOf[jakarta.persistence.Table])
      if (table != null && table.name().nonEmpty) {
        table.name()
      } else {
        entity.getName
      }
    }.getOrElse(entity.getName)
  }

  private def queryOverview(): QueryResult = {
    entityManagerFactory match {
      case None => QueryResult("jpa:overview", List.empty, 0, 0, 0)
      case Some(emf) =>
        val em = emf.createEntityManager()
        try {
          val entities = emf.getMetamodel.getEntities.asScala

          val data = entities.flatMap { entity =>
            val entityName = entity.getName
            Try {
              val count = getEntityCount(em, entityName)
              Map[String, AnyRef](
                "entityName" -> entityName,
                "javaType" -> entity.getJavaType.getName,
                "tableName" -> getTableName(entity),
                "recordCount" -> Long.box(count),
                "attributes" -> Int.box(entity.getAttributes.size())
              )
            }.recover {
              case e: Exception =>
                logger.debug(s"Error querying entity $entityName: ${e.getMessage}")
                null
            }.toOption
          }.filter(_ != null).toList

          QueryResult(
            dataSourceId = "jpa:overview",
            data = data,
            totalCount = data.size,
            limit = data.size,
            offset = 0
          )
        } finally {
          em.close()
        }
    }
  }

  private def queryEntity(entityName: String, filters: Map[String, AnyRef], limit: Int, offset: Int): QueryResult = {
    entityManagerFactory match {
      case None => QueryResult(s"jpa:entity:$entityName", List.empty, 0, limit, offset)
      case Some(emf) =>
        val em = emf.createEntityManager()
        try {
          // Get total count
          val totalCount = getEntityCount(em, entityName)

          // Build query with filters
          val jpql = new StringBuilder(s"SELECT e FROM $entityName e")

          if (filters.nonEmpty) {
            jpql.append(" WHERE ")
            val conditions = filters.keys.map(key => s"e.$key = :$key").mkString(" AND ")
            jpql.append(conditions)
          }

          val query: TypedQuery[AnyRef] = em.createQuery(jpql.toString(), classOf[AnyRef])

          // Set filter parameters
          filters.foreach { case (key, value) =>
            query.setParameter(key, value)
          }

          // Apply pagination
          query.setFirstResult(offset)
          query.setMaxResults(limit)

          val results = query.getResultList.asScala.toList

          // Convert entities to maps
          val data = results.map(convertEntityToMap)

          QueryResult(
            dataSourceId = s"jpa:entity:$entityName",
            data = data,
            totalCount = totalCount,
            limit = limit,
            offset = offset,
            stats = Map[String, AnyRef]("entityName" -> entityName)
          )
        } catch {
          case e: Exception =>
            logger.error(s"Error querying entity $entityName: ${e.getMessage}", e)
            QueryResult(
              dataSourceId = s"jpa:entity:$entityName",
              data = List.empty,
              totalCount = 0,
              limit = limit,
              offset = offset
            )
        } finally {
          em.close()
        }
    }
  }

  private def convertEntityToMap(entity: AnyRef): Map[String, AnyRef] = {
    Try {
      // Use ObjectMapper to convert entity to map
      val javaMap = objectMapper.convertValue(entity, classOf[java.util.Map[String, AnyRef]])
      javaMap.asScala.toMap
    }.recover {
      case e: Exception =>
        logger.debug(s"Could not convert entity to map: ${e.getMessage}")
        Map[String, AnyRef](
          "type" -> entity.getClass.getSimpleName,
          "value" -> entity.toString
        )
    }.get
  }
}
